### PR TITLE
i150 cxx headers

### PR DESCRIPTION
**Pull request by kubanrob**
See issue #150 for details.

Init apps using OpenMP should be updated in order to link properly.

### CODE REVIEW

File: kernel/runtime/cxx-headers/mcconf.module [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+			"forward_list",
+			"set",
+			"map",
+			"unordered_set",
+			"unordered_map",
+			"stack",
+			"queue",
+			"iterator",
+			"algorithm",
+			"# Numerics library",
```
Comment (gypsephi) [inline kernel/runtime/cxx-headers/mcconf.module:53]: is that right?
is '# Numerics library' a real include?

**Comment by gypsephi** [comment]
mcconf fails for IHK:
./3rdparty/mcconf/mcconf -i kernel-ihk.config
processing configuration /home/pgypser/mythos_robert/kernel-ihk.config
searching modules in /home/pgypser/mythos_robert/./3rdparty/mcconf/stdmodules
searching modules in /home/pgypser/mythos_robert/kernel
searching modules in /home/pgypser/mythos_robert/arch-config
added modules to resolve dependencies: amd64-cpu-fpu, amd64-cpu-regs, apboot-common, apboot-ihk, app-common, app-mlog, assert, async-mutex-delegating, boot-init-loader-amd64, capability-spinning, capability-utils, clflush-amd64, core-local-memory, cxx-globals, cxx-headers-app, cxxabi-app, cxxabi-kernel, datatypes, error-trace, fnv-1a, gcc-kernel-pc, host-info-table-knc, kernel-bug-amd64, kernel-entry-amd64, kernel-idle-hlt, kernel-main, kernel-threadid, kmem-common, local-x2apic, mcconf-makefile, memory-root, mlog-async, mlog-base, monitor-common, mstring, mythos-caps, mythos-error, mythos-init, mythos-invocationBuffer, mythos-pci-mpsc-queue, nested-monitor-delegating, objects-capmap, objects-common, objects-cpudriver-knc, objects-execution-context, objects-interrupt-control, objects-kernel-memory, objects-memory-amd64, objects-portal, objects-rapl-driver-intel, objects-scheduling-context, objects_example, optional, ostream, pic-disable-x86, posix-headers-app, runtime-async, runtime-kobject, runtime-memory, runtime-mlog, segments-amd64, segments-gdt-amd64, simple-monitor-home, stacktrace, synchronous-task, thread-pause-amd64, user-mutex-tidex, util-bitfield, util-castable, util-elf64, util-events, util-first-fit-heap-intrusive, util-linked-list, util-mutex-tidex, utils-alignment
unresolved dependency bits/stl_algobase.h required by [cxxabi-kernel] provided by [gcc-host-pc]

**Comment by kubanrob** [comment]
> We should have this separate from the mcconf module dependencies because these cover more stuff and implicit dependencies as well. Is that even possible?

I don't think that is necessary and would introduce yet another concept into the module description.

The main problem with topology ordering and the implicit dependencies would be cycles, but I think we can treat the modules in a connected component (dt. Zusammenhangskomponente) as unordered in respect to each other ... and maybe order them by name.

Then we would probably need some additional diagnostic to hunt errors in the dependencies. We could mark the connected components in the GraphVis dependency plot by putting them into clusters.

**Comment by rottaran** [comment]
Okay, now I see the underlying challenge: the userland and kernel subprojects depend on different variants of the same header file name. This would produce no conflict in the build because both could see different search pathes to their respective variant. However mcconf things that each header file name can be provided at most once independently of `-I` search pathes.

Parameters should not be aggregated into a single global set of configuration parameters. Instead, in order to build a file, the parameters are aggregated per target file based on what the respective source file actually depends on.
But then, we loose the convex hull operation that we use now to resolve dependencies.

Splitting into separate configurations for subprojects would be more helpful. And handling external libraries differently.
Both nothing quick to fix.

**Review by rottaran — APPROVED** [review]

**Pull request merged**